File: momentra/momentra/Shell/GroupActive/Shared/GroupExperienceChecklistUi.swift
import SwiftUI

/// Quick Add checklist sheet: title + category chips + seed packing list.
struct ExperienceChecklistBody: View {
    var momentId: String?
    var onDismiss: () -> Void
    var onSaved: () -> Void
    var accent: SheetAccent = SheetAccent(
        accent: Color(hex: "#14B8A6"),
        accentEnd: Color(hex: "#0F766E"),
        soft: Color(hex: "#14B8A6").opacity(0.2)
    )
    /// When set, sheet edits an existing checklist item instead of creating.
    var editingItem: GroupPlanningItem? = nil

    @State private var title = ""
    @State private var selectedCategoryLabel = GroupExperienceChecklistCatalog.defaultLabel()
    @State private var submitting = false
    @State private var seeding = false
    @State private var deleting = false
    @State private var showDeleteConfirm = false
    @State private var error: String?
    @State private var didPrefill = false

    private var isEditing: Bool { editingItem?.planningItemId != nil }

    private var categoryCode: String {
        GroupExperienceChecklistCatalog.categories
            .first(where: { $0.label == selectedCategoryLabel })?.code
            ?? GroupExperienceChecklistCatalog.defaultCode()
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 16) {
            SheetHeader(
                icon: "checklist",
                title: isEditing ? "Edit checklist" : "Checklist",
                subtitle: isEditing ? "Update packing & essentials" : "Shared packing & essentials",
                accent: accent
            )

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "Item")
                SheetField(value: $title, placeholder: "What to pack or prepare", minHeight: 42)
            }

            VStack(alignment: .leading, spacing: 8) {
                FieldLabel(text: "Category")
                ChipRow(
                    options: GroupExperienceChecklistCatalog.categories.map(\.label),
                    selected: $selectedCategoryLabel,
                    accent: accent
                )
            }

            if let error {
                Text(error)
                    .font(.plusJakarta(size: 12))
                    .foregroundStyle(Color(hex: "#F87171"))
            }

            PrimaryCta(
                label: isEditing ? "Save" : "Add",
                enabled: !(momentId ?? "").isEmpty
                    && !title.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                    && !submitting
                    && !seeding
                    && !deleting,
                accent: accent,
                loading: submitting
            ) {
                Task { await saveItem() }
            }

            if isEditing {
                Button {
                    showDeleteConfirm = true
                } label: {
                    Text(deleting ? "Deleting…" : "Delete from this moment")
                        .font(.plusJakarta(size: 14, weight: .semibold))
                        .foregroundStyle(Color(hex: "#F87171"))
                        .frame(maxWidth: .infinity, alignment: .leading)
                        .padding(.vertical, 8)
                }
                .buttonStyle(.plain)
                .disabled(submitting || deleting || (momentId ?? "").isEmpty)
            } else {
                PrimaryCta(
                    label: seeding ? "Seeding…" : "Seed packing list",
                    enabled: !(momentId ?? "").isEmpty && !submitting && !seeding,
                    accent: accent,
                    loading: seeding,
                    lightLabel: true
                ) {
                    Task { await seedPackingList() }
                }
            }
        }
        .onAppear { prefillIfNeeded() }
        .confirmationDialog(
            "Delete this checklist item?",
            isPresented: $showDeleteConfirm,
            titleVisibility: .visible
        ) {
            Button("Delete", role: .destructive) {
                Task { await deleteItem() }
            }
            Button("Cancel", role: .cancel) {}
        } message: {
            Text("Removes it from this moment only. The shared packing seed list is unchanged.")
        }
    }

    private func prefillIfNeeded() {
        guard !didPrefill, let item = editingItem else { return }
        didPrefill = true
        title = item.title ?? ""
        if let code = item.categoryCode, GroupExperienceChecklistCatalog.isChecklistCode(code) {
            selectedCategoryLabel = GroupExperienceChecklistCatalog.label(forCode: code)
        }
    }

    @MainActor
    private func saveItem() async {
        guard let momentId, !momentId.isEmpty else { return }
        let trimmed = title.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return }
        submitting = true
        error = nil
        defer { submitting = false }
        do {
            if let planningItemId = editingItem?.planningItemId {
                _ = try await APIClient.shared.updatePlanningItem(
                    momentId: momentId,
                    planningItemId: planningItemId,
                    title: trimmed,
                    categoryCode: categoryCode,
                    status: editingItem?.status
                )
            } else {
                _ = try await APIClient.shared.createPlanningItem(
                    momentId: momentId,
                    title: trimmed,
                    categoryCode: categoryCode
                )
            }
            onSaved()
            onDismiss()
        } catch {
            self.error = error.localizedDescription
        }
    }

    @MainActor
    private func deleteItem() async {
        guard let momentId, !momentId.isEmpty,
              let planningItemId = editingItem?.planningItemId
        else { return }
        deleting = true
        error = nil
        defer { deleting = false }
        do {
            _ = try await APIClient.shared.deletePlanningItem(
                momentId: momentId,
                planningItemId: planningItemId
            )
            onSaved()
            onDismiss()
        } catch {
            self.error = error.localizedDescription
        }
    }

    @MainActor
    private func seedPackingList() async {
        guard let momentId, !momentId.isEmpty else { return }
        seeding = true
        error = nil
        defer { seeding = false }
        do {
            let existing = (try? await APIClient.shared.listPlanningItems(momentId: momentId))?.items ?? []
            var existingKeys = Set(
                existing
                    .filter { GroupExperienceChecklistCatalog.isChecklistCode($0.categoryCode) }
                    .compactMap { item -> String? in
                        guard let t = item.title?.trimmingCharacters(in: .whitespacesAndNewlines).lowercased(),
                              let c = item.categoryCode?.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
                        else { return nil }
                        return "\(c)|\(t)"
                    }
            )
            var created = 0
            for seed in GroupExperienceChecklistCatalog.seedPackingList {
                let key = "\(seed.code)|\(seed.title.trimmingCharacters(in: .whitespacesAndNewlines).lowercased())"
                if existingKeys.contains(key) { continue }
                _ = try await APIClient.shared.createPlanningItem(
                    momentId: momentId,
                    title: seed.title,
                    categoryCode: seed.code
                )
                existingKeys.insert(key)
                created += 1
            }
            onSaved()
            if created > 0 {
                onDismiss()
            } else {
                error = "Packing list already seeded"
            }
        } catch {
            self.error = error.localizedDescription
        }
    }
}

/// Moments Checklist section grouped by category with DONE ↔ OPEN toggle.
struct MomentsChecklistSection: View {
    let items: [GroupPlanningItem]
    var momentId: String?
    var chrome: MomentsChrome
    var onChanged: () -> Void = {}
    var onAdd: (() -> Void)? = nil

    @State private var togglingId: String?
    /// Category codes that are collapsed. Empty = all expanded (least surprise).
    @State private var collapsedCodes: Set<String> = []
    @State private var editingItem: GroupPlanningItem?

    private var groups: [(category: GroupExperienceChecklistCatalog.Category, items: [GroupPlanningItem])] {
        GroupExperienceChecklistCatalog.groupedByCategory(items)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 12) {
            MomentsSectionHeader(title: "Checklist  ✅", chrome: chrome)

            if groups.isEmpty {
                GroupEmptySection(
                    message: "No checklist items yet",
                    detail: "Add essentials from Quick Add, or seed the packing list."
                )
                if let onAdd {
                    Button("Add checklist item", action: onAdd)
                        .font(.plusJakarta(size: 12, weight: .semibold))
                        .foregroundStyle(chrome.accent)
                        .buttonStyle(.plain)
                }
            } else {
                ForEach(groups, id: \.category.code) { group in
                    let expanded = !collapsedCodes.contains(group.category.code)
                    let doneCount = group.items.filter {
                        ($0.status ?? "").caseInsensitiveCompare("DONE") == .orderedSame
                    }.count
                    VStack(alignment: .leading, spacing: 6) {
                        Button {
                            if expanded {
                                collapsedCodes.insert(group.category.code)
                            } else {
                                collapsedCodes.remove(group.category.code)
                            }
                        } label: {
                            HStack(spacing: 6) {
                                Image(systemName: expanded ? "chevron.down" : "chevron.right")
                                    .font(.system(size: 11, weight: .bold))
                                    .foregroundStyle(chrome.accent)
                                Text(group.category.label)
                                    .font(.plusJakarta(size: 12, weight: .bold))
                                    .foregroundStyle(chrome.accent)
                                Spacer(minLength: 0)
                                Text("\(doneCount)/\(group.items.count)")
                                    .font(.plusJakarta(size: 11, weight: .semibold))
                                    .foregroundStyle(chrome.secondary)
                            }
                        }
                        .buttonStyle(.plain)

                        if expanded {
                            ForEach(Array(group.items.enumerated()), id: \.offset) { _, item in
                                checklistRow(item)
                            }
                        }
                    }
                }
                if let onAdd {
                    Button("Add item", action: onAdd)
                        .font(.plusJakarta(size: 12, weight: .semibold))
                        .foregroundStyle(chrome.accent)
                        .buttonStyle(.plain)
                }
            }
        }
        .sheet(isPresented: Binding(
            get: { editingItem != nil },
            set: { if !$0 { editingItem = nil } }
        )) {
            if let item = editingItem {
                NativeSheetScaffold(
                    title: "Edit checklist",
                    onClose: { editingItem = nil },
                    background: chrome.card
                ) {
                    ScrollView {
                        ExperienceChecklistBody(
                            momentId: momentId,
                            onDismiss: { editingItem = nil },
                            onSaved: {
                                editingItem = nil
                                onChanged()
                            },
                            accent: SheetAccent(
                                accent: chrome.accent,
                                accentEnd: chrome.accent,
                                soft: chrome.accent.opacity(0.2)
                            ),
                            editingItem: item
                        )
                        .padding(.horizontal, 20)
                        .padding(.top, 16)
                        .padding(.bottom, 28)
                    }
                }
                .presentationDetents([.medium, .large])
                .presentationDragIndicator(.visible)
            }
        }
    }

    @ViewBuilder
    private func checklistRow(_ item: GroupPlanningItem) -> some View {
        let done = (item.status ?? "").caseInsensitiveCompare("DONE") == .orderedSame
        let id = item.planningItemId
        HStack(spacing: 4) {
            Button {
                guard let id, let momentId, !momentId.isEmpty, togglingId == nil else { return }
                let next = done ? "OPEN" : "DONE"
                Task { await toggle(item: item, id: id, momentId: momentId, status: next) }
            } label: {
                Image(systemName: done ? "checkmark.square.fill" : "square")
                    .font(.system(size: 20))
                    .foregroundStyle(done ? chrome.accent : chrome.secondary)
            }
            .buttonStyle(.plain)
            .disabled(id == nil || momentId == nil || togglingId == id)

            Text(item.title ?? "Item")
                .font(.plusJakarta(size: 13, weight: .medium))
                .foregroundStyle(done ? chrome.secondary : chrome.text)
                .strikethrough(done)
            Spacer(minLength: 0)

            if id != nil, momentId != nil {
                Button {
                    editingItem = item
                } label: {
                    Text("Edit")
                        .font(.plusJakarta(size: 12, weight: .semibold))
                        .foregroundStyle(chrome.accent)
                        .padding(.horizontal, 8)
                        .padding(.vertical, 4)
                }
                .buttonStyle(.plain)
                .accessibilityLabel("Edit checklist item")
            }
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 8)
        .background(chrome.card)
        .overlay(RoundedRectangle(cornerRadius: 12).stroke(chrome.border, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: 12))
    }

    @MainActor
    private func toggle(item: GroupPlanningItem, id: String, momentId: String, status: String) async {
        togglingId = id
        defer { togglingId = nil }
        do {
            _ = try await APIClient.shared.updatePlanningItem(
                momentId: momentId,
                planningItemId: id,
                title: item.title ?? "Item",
                categoryCode: item.categoryCode,
                status: status
            )
            onChanged()
        } catch {
            // Keep prior UI state; list refresh will reconcile.
        }
    }
}

File: apk/app/src/main/java/com/example/momentra/ui/shell/group/shared/GroupExperienceChecklistUi.kt
package com.example.momentra.ui.shell.group.shared

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.Checkbox
import androidx.compose.material3.CheckboxDefaults
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.example.momentra.R
import com.example.momentra.data.api.GroupLifePlanningItemDto
import com.example.momentra.data.repository.GroupSliceRepository
import com.example.momentra.ui.shell.components.MomentraModalBottomSheet
import com.example.momentra.ui.shell.group.wedding.create.ChipRow
import com.example.momentra.ui.shell.group.wedding.create.FieldLabel
import com.example.momentra.ui.shell.group.wedding.create.PrimaryCta
import com.example.momentra.ui.shell.group.wedding.create.SheetAccent
import com.example.momentra.ui.shell.group.wedding.create.SheetField
import com.example.momentra.ui.shell.group.wedding.create.SheetHeader
import com.example.momentra.ui.theme.PlusJakartaSans
import kotlinx.coroutines.launch
import java.util.Locale

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ExperienceChecklistAddSheet(
    visible: Boolean,
    momentId: String?,
    onDismiss: () -> Unit,
    onSaved: () -> Unit,
    accent: SheetAccent,
    surface: Color = Color(0xFF1C1A24),
    repository: GroupSliceRepository = remember { GroupSliceRepository() },
) {
    if (!visible) return
    MomentraModalBottomSheet(
        onDismissRequest = onDismiss,
        containerColor = surface,
        skipPartiallyExpanded = false,
        dragHandle = {
            Box(
                modifier = Modifier
                    .padding(top = 12.dp, bottom = 4.dp)
                    .size(width = 40.dp, height = 5.dp)
                    .clip(RoundedCornerShape(100.dp))
                    .background(Color.White.copy(alpha = 0.2f)),
            )
        },
    ) {
        ExperienceChecklistSheetBody(
            momentId = momentId,
            repository = repository,
            onDismiss = onDismiss,
            onSaved = onSaved,
            accent = accent,
        )
    }
}

/** Quick Add / Moments sheet body: title + category chips + seed packing list. */
@Composable
fun ExperienceChecklistSheetBody(
    momentId: String?,
    repository: GroupSliceRepository,
    onDismiss: () -> Unit,
    onSaved: () -> Unit,
    accent: SheetAccent,
) {
    var title by remember { mutableStateOf("") }
    var categoryCode by remember { mutableStateOf(GroupExperienceChecklistCatalog.defaultCode()) }
    var submitting by remember { mutableStateOf(false) }
    var seeding by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val scope = rememberCoroutineScope()
    val categoryLabels = GroupExperienceChecklistCatalog.categories.map { it.label }
    val selectedLabel = GroupExperienceChecklistCatalog.labelForCode(categoryCode)

    SheetHeader(
        R.drawable.ic_group_qa_calendar,
        "Checklist",
        "Shared packing & essentials",
        accent = accent,
    )
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("Item")
        SheetField(title, { title = it }, "What to pack or prepare", minHeight = 42)
    }
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        FieldLabel("Category")
        ChipRow(categoryLabels, selectedLabel, accent) { label ->
            categoryCode = GroupExperienceChecklistCatalog.categories
                .firstOrNull { it.label == label }?.code
                ?: GroupExperienceChecklistCatalog.defaultCode()
        }
    }
    error?.let {
        Text(it, color = Color(0xFFF87171), fontSize = 12.sp, fontFamily = PlusJakartaSans)
    }
    PrimaryCta(
        label = "Add",
        enabled = !momentId.isNullOrBlank() && title.isNotBlank() && !submitting && !seeding,
        accent = accent,
        loading = submitting,
        onClick = {
            val mid = momentId ?: return@PrimaryCta
            val trimmed = title.trim()
            if (trimmed.isEmpty()) return@PrimaryCta
            scope.launch {
                submitting = true
                error = null
                repository.createPlanningItem(
                    momentId = mid,
                    title = trimmed,
                    categoryCode = categoryCode,
                ).onSuccess {
                    submitting = false
                    onSaved()
                    onDismiss()
                }.onFailure { e ->
                    submitting = false
                    error = e.message ?: "Could not add checklist item"
                }
            }
        },
    )
    PrimaryCta(
        label = if (seeding) "Seeding…" else "Seed packing list",
        enabled = !momentId.isNullOrBlank() && !submitting && !seeding,
        accent = accent,
        loading = seeding,
        lightLabel = true,
        onClick = {
            val mid = momentId ?: return@PrimaryCta
            scope.launch {
                seeding = true
                error = null
                val existing = repository.listPlanningItems(mid).getOrNull()?.items.orEmpty()
                val existingKeys = existing
                    .filter { GroupExperienceChecklistCatalog.isChecklistCode(it.categoryCode) }
                    .mapNotNull { item ->
                        val t = item.title?.trim()?.lowercase(Locale.US) ?: return@mapNotNull null
                        val c = item.categoryCode?.trim()?.uppercase(Locale.US) ?: return@mapNotNull null
                        "$c|$t"
                    }
                    .toMutableSet()
                var created = 0
                var failed: String? = null
                for ((code, seedTitle) in GroupExperienceChecklistCatalog.seedPackingList) {
                    val key = "${code}|${seedTitle.trim().lowercase(Locale.US)}"
                    if (key in existingKeys) continue
                    val result = repository.createPlanningItem(
                        momentId = mid,
                        title = seedTitle,
                        categoryCode = code,
                    )
                    if (result.isSuccess) {
                        existingKeys.add(key)
                        created++
                    } else {
                        failed = result.exceptionOrNull()?.message
                        break
                    }
                }
                seeding = false
                if (failed != null) {
                    error = failed
                } else {
                    onSaved()
                    if (created > 0) onDismiss()
                    else error = "Packing list already seeded"
                }
            }
        },
    )
}

@Composable
fun MomentsChecklistSection(
    items: List<GroupLifePlanningItemDto>,
    momentId: String?,
    chrome: MomentsChrome,
    repository: GroupSliceRepository = remember { GroupSliceRepository() },
    onChanged: () -> Unit = {},
    onAdd: (() -> Unit)? = null,
) {
    val groups = remember(items) { GroupExperienceChecklistCatalog.groupedByCategory(items) }
    val scope = rememberCoroutineScope()
    var togglingId by remember { mutableStateOf<String?>(null) }

    MomentsSectionHeader("Checklist  ✅", chrome)
    if (groups.isEmpty()) {
        GroupEmptySection(
            "No checklist items yet",
            "Add essentials from Quick Add, or seed the packing list.",
        )
        if (onAdd != null) {
            Text(
                "Add checklist item",
                color = chrome.accent,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                fontFamily = PlusJakartaSans,
                modifier = Modifier
                    .clickable(onClick = onAdd)
                    .padding(vertical = 4.dp),
            )
        }
        return
    }

    Column(verticalArrangement = Arrangement.spacedBy(12.dp)) {
        groups.forEach { (category, groupItems) ->
            Column(verticalArrangement = Arrangement.spacedBy(6.dp)) {
                Text(
                    category.label,
                    color = chrome.accent,
                    fontSize = 12.sp,
                    fontWeight = FontWeight.Bold,
                    fontFamily = PlusJakartaSans,
                )
                groupItems.forEach { item ->
                    val done = item.status?.equals("DONE", ignoreCase = true) == true
                    val id = item.planningItemId
                    Row(
                        modifier = Modifier
                            .fillMaxWidth()
                            .clip(RoundedCornerShape(12.dp))
                            .background(chrome.card)
                            .border(1.dp, chrome.border, RoundedCornerShape(12.dp))
                            .padding(horizontal = 8.dp, vertical = 4.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                    ) {
                        Checkbox(
                            checked = done,
                            onCheckedChange = { checked ->
                                if (id.isNullOrBlank() || momentId.isNullOrBlank() || togglingId != null) return@Checkbox
                                val next = if (checked) "DONE" else "OPEN"
                                scope.launch {
                                    togglingId = id
                                    repository.updatePlanningItem(
                                        momentId = momentId,
                                        planningItemId = id,
                                        title = item.title ?: "Item",
                                        categoryCode = item.categoryCode,
                                        status = next,
                                    ).onSuccess { onChanged() }
                                    togglingId = null
                                }
                            },
                            enabled = !id.isNullOrBlank() && !momentId.isNullOrBlank() && togglingId != id,
                            colors = CheckboxDefaults.colors(
                                checkedColor = chrome.accent,
                                uncheckedColor = chrome.secondary,
                                checkmarkColor = Color.White,
                            ),
                        )
                        Text(
                            item.title ?: "Item",
                            color = if (done) chrome.secondary else chrome.text,
                            fontSize = 13.sp,
                            fontWeight = FontWeight.Medium,
                            fontFamily = PlusJakartaSans,
                            textDecoration = if (done) TextDecoration.LineThrough else null,
                            modifier = Modifier.weight(1f),
                        )
                    }
                }
            }
        }
        if (onAdd != null) {
            Text(
                "Add item",
                color = chrome.accent,
                fontSize = 12.sp,
                fontWeight = FontWeight.SemiBold,
                fontFamily = PlusJakartaSans,
                modifier = Modifier
                    .clickable(onClick = onAdd)
                    .padding(vertical = 2.dp),
            )
        }
    }
}
